### PR TITLE
tests: test handling normal entries

### PR DESCRIPTION
**Pull request by BusyJay**
This pr add two tests that test applying the normal entries and basic applying flow. Admin entries will be added in the next pr.

@siddontang @hhkbp2 PTAL

### CODE REVIEW

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -1161,3 +1181,339 @@ impl Runnable<Task> for Runner {
...
+        let (resp_tx, resp_rx) = mpsc::channel();
+        runner.run(Task::propose(1,
+                                 1,
+                                 Uuid::new_v4(),
+                                 false,
+                                 0,
+                                 box move |resp| {
+                                     resp_tx.send(resp).unwrap();
+                                 }));
+        // non registered region should be ignored and notify failed.
```
Comment (andelf) [inline src/raftstore/store/worker/apply.rs]: unregistered

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -1161,3 +1181,339 @@ impl Runnable<Task> for Runner {
...
+    use tempdir::TempDir;
+    use rocksdb::DB;
+    use protobuf::Message;
+    use kvproto::metapb::RegionEpoch;
+    use kvproto::raft_cmdpb::CmdType;
+
+    use super::*;
+    use storage::{CF_WRITE, ALL_CFS};
+
+    pub fn create_tmp_engine(label: &str) -> (TempDir, Arc<DB>) {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: May path be better?

Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: No, path will be deleted once `TempDir` is dropped.

Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: sorry, what I mean is renaming label to path.

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -1161,3 +1181,339 @@ impl Runnable<Task> for Runner {
...
+                                 0,
+                                 box move |resp| {
+                                     resp_tx.send(resp).unwrap();
+                                 }));
+        // non registered region should be ignored and notify failed.
+        assert!(rx.try_recv().is_err());
+        let resp = resp_rx.try_recv().unwrap();
+        assert!(resp.get_header().get_error().has_region_not_found());
+
+        let normal_uuid = Uuid::new_v4();
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: here is normal, do we have any un-normal ?

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: src/raftstore/store/worker/apply.rs [hunk 4]
```diff
@@ -1161,3 +1181,339 @@ impl Runnable<Task> for Runner {
...
+            let mut entry = Entry::new();
+            entry.set_index(index);
+            entry.set_term(term);
+            EntryBuilder {
+                entry: entry,
+                req: req,
+            }
+        }
+
+        fn capture_resp(self,
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: seem this function doesn't change the Builder, is it proper here?

**Comment by siddontang** [comment]
Greate!

Rest LGTM

**Comment by siddontang** [comment]
CI failed

```
thread 'raftstore::test_stats::test_server_store_snap_stats' panicked at 'must detect snapshot sending/receiving', tests/raftstore/test_stats.rs:147
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: tikv::util::panic_hook::track_hook::{{closure}}
   1: <std::thread::local::LocalKey<T>>::with
   2: tikv::util::panic_hook::track_hook
   3: core::ops::Fn::call
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: tests::raftstore::test_stats::must_detect_snap
   7: tests::raftstore::test_stats::test_server_store_snap_stats
   8: <F as test::FnBox<T>>::call_box
   9: std::panicking::try::do_call
  10: __rust_maybe_catch_panic
  11: std::panicking::try::do_call
  12: __rust_maybe_catch_panic
  13: <F as alloc::boxed::FnBox<A>>::call_box
  14: std::sys::imp::thread::Thread::new::thread_start
  15: _pthread_body
  16: _pthread_start
test raftstore::test_stats::test_server_store_snap_stats ... FAILED
```

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: src/raftstore/store/worker/apply.rs [hunk 5]
```diff
@@ -1161,3 +1181,339 @@ impl Runnable<Task> for Runner {
...
+        let written_keys = delegate.metrics.written_keys;
+        let size_diff_hint = delegate.metrics.size_diff_hint;
+        let put_entry = EntryBuilder::new(2, 2).put_cf(CF_LOCK, b"k1", b"v1").epoch(1, 3).build();
+        delegate.handle_raft_entry_normal(put_entry);
+        let lock_handle = db.cf_handle(CF_LOCK).unwrap();
+        assert_eq!(db.get_cf(lock_handle, &dk_k1).unwrap().unwrap(), b"v1");
+        assert_eq!(delegate.metrics.lock_cf_written_bytes,
+                   lock_written_bytes + 5);
+        assert!(delegate.metrics.written_bytes >= written_bytes + 5);
+        assert_eq!(delegate.metrics.written_keys, written_keys + 2);
```
Comment (hhkbp2) [inline src/raftstore/store/worker/apply.rs]: Why is it 5 and 2 here?

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/raftstore/store/worker/apply.rs (hunk 5 above)
Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: KVs written are ("zk1", "v1") and (applied_state_key, applied_state). And written_bytes should be larger than the length of all key values.

**Comment by hhkbp2** [comment]
LGTM

**Comment by siddontang** [comment]
CI still failed 

@BusyJay

**Comment by siddontang** [comment]
@BusyJay 

```
    --> src/raftstore/store/worker/apply.rs:1235:24
     |
1235 |             delegates: HashMap::new(),
     |                        ^^^^^^^^^^^^^^ expected struct `ordermap::OrderMap`, found struct `std::collections::HashMap`
     |
     = note: expected type `ordermap::OrderMap<u64, raftstore::store::worker::apply::ApplyDelegate>`
                found type `std::collections::HashMap<_, _>`

```

Please fix CI.

**Pull request merged**